pkg/registry/core/pod/storage/eviction.go

```diff
@@ -324,11 +338,23 @@ func addConditionAndDeletePod(r *EvictionREST, ctx context.Context, name string,
 			return podObj, nil
 		}
 
-		podCopyUpdated := rest.DefaultUpdatedObjectInfo(pod, conditionAppender)
+		// obj will be obtained in setPreconditions transformer
+		podUpdatedObjectInfo := rest.DefaultUpdatedObjectInfo(nil, setPreconditions, conditionAppender) // order important
 
-		if _, _, err = r.store.Update(ctx, name, podCopyUpdated, rest.ValidateAllObjectFunc, rest.ValidateAllObjectUpdateFunc, false, &metav1.UpdateOptions{}); err != nil {
+		updatedPodObject, _, err := r.store.Update(ctx, name, podUpdatedObjectInfo, rest.ValidateAllObjectFunc, rest.ValidateAllObjectUpdateFunc, false, &metav1.UpdateOptions{})
+		if err != nil {
```

Comment:
I did some more testing, and a bit problematic part here is that we are getting different errors now:

when trying to evict pod with `.deleteOptions.preconditions.uid`

before we were getting: `Error from server (Conflict): Operation cannot be fulfilled on Pod "nginx": the UID in the precondition (596f4808-3fd2-4de3-b028-4980faa3ba29) does not match the UID in record (7d9eec6a-0e46-498f-94c3-578299306a31). The object might have been deleted and then recreated`

now we get: `The Pod "nginx" is invalid: metadata.uid: Invalid value: "596f4808-3fd2-4de3-b028-4980faa3ba29": field is immutable`

which could be confusing to get during an eviction request.

It is much better when using `.deleteOptions.preconditions.resourceVersion`

before we got: `Error from server (Conflict): Operation cannot be fulfilled on Pod "nginx": the ResourceVersion in the precondition (469) does not match the ResourceVersion in record (522539). The object might have been modified`

now we are getting: `Error from server (Conflict): Operation cannot be fulfilled on pods "nginx": the object has been modified; please apply your changes to the latest version and try again`

This one is not that confusing but the before message was more appropriate.

We could try to reintroduce the preconditions, or do some processing/validation after when we get an error and generate our messsages (at least for the uid one).

thoughts?

Reply:
rather than setting the uid / resourceVersion into podWithPreconditions, we can do the comparison directly and return the conflict errors we want that exactly match the delete uid/rv errors:

```go
podWithPreconditions := oldObj.(*api.Pod).DeepCopy()
if uid := options.Preconditions.UID; uid != nil && len(*uid) > 0 && *uid != podWithPreconditions.UID {
  return nil, errors.NewConflict(
    schema.GroupResource{Group: "", Resource: "Pod"},
    podWithPreconditions.Name,
    fmt.Errorf("the UID in the precondition (%s) does not match the UID in record (%s). The object might have been deleted and then recreated", *uid, podWithPreconditions.UID),
  )
}
if rv := options.Preconditions.ResourceVersion; rv != nil && len(*rv) > 0 && *rv != podWithPreconditions.ResourceVersion {
  return nil, errors.NewConflict(
    schema.GroupResource{Group: "", Resource: "Pod"},
    podWithPreconditions.Name,
    fmt.Errorf("the ResourceVersion in the precondition (%s) does not match the ResourceVersion in record (%s). The object might have been modified", *rv, podWithPreconditions.ResourceVersion),
  )
}
return podWithPreconditions, nil
```

Reply:
thanks for checking the user-facing behavior in conflicts

Reply:
ah, we can just return the error from the transformer ^^. Works well now. Thanks